### PR TITLE
chore(v2): added minio and s3 pipeline root support in v2 launcher

### PR DESCRIPTION
**Pull request by capri-xiyue**
**Description of your changes:**

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by capri-xiyue** [comment]
> Thank you, I think this mostly look good.
> 
> One nice thing to have is adding a sample that runs on minio in the same PR.
> You can follow documentation in https://github.com/kubeflow/pipelines/tree/master/v2/test.
> 
> (Note, I plan to do some refactoring of v2 sample test, so I'd recommend waiting for a few days before trying that)

Will try to add sample test for minio in this PR.

**Comment by Ark-kun** [comment]
I'm not sure we should add any cloud-specific information like AWS-specific environment variables to all compiled pipelines.

I also see that the added YAML is constant so it is redundant. If you think it needs to be added, please add it in the backend transformer here: https://github.com/kubeflow/pipelines/blob/d9c019641ef9ebd78db60cdb78ea29b0d9933008/backend/src/apiserver/resource/resource_manager.go#L396

**Comment by Bobgy** [comment]
@Ark-kun I think both @capri-xiyue and I agree to that, this PR is just an initial version to get minio working e2e in v2 compatible mode.

V2 compatible version will take a while to be released, so we think it's fine to switch to a backend implementation later, but probably before release, what do you think?

**Comment by Bobgy** [comment]
@capri-xiyue I recalled that, one prerequisite to make this work is setting up a kfp-s3-secret in the KFP ci test cluster. Let me provide an example of how to do that

**Comment by capri-xiyue** [comment]
> @capri-xiyue I recalled that, one prerequisite to make this work is setting up a kfp-s3-secret in the KFP ci test cluster. Let me provide an example of how to do that

I submitted a pr for it https://github.com/kubeflow/testing/pull/947. I'm not sure whether I did it correctly. Feel free to leave a comment in that pr.

**Comment by Bobgy** [comment]
I just checked the run result in test infra, looks great!
https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/6842e5e8-df1b-456b-a358-c74571007cdf

Would URIs be more informative if we keep the parameters including endpoint? e.g. if users have multiple different minio instances, they can use the endpoint to distinguish minio URIs. I think for stuff we store in MLMD, we want them to be globally unique identifiers.
image

File: v2/component/launcher.go [hunk 1]
```diff
@@ -63,13 +64,21 @@ type bucketConfig struct {
...
-		u = fmt.Sprintf("%s?prefix=%s", u, b.prefix)
+	if len(b.prefix) > 0 && len(b.queryString) > 0 {
+		u = fmt.Sprintf("%s%s&prefix=%s", u, b.queryString, b.prefix)
+	} else {
+		if len(b.prefix) > 0 {
+			u = fmt.Sprintf("%s?prefix=%s", u, b.prefix)
+		}
+		if len(b.queryString) > 0 {
+			u = fmt.Sprintf("%s%s", u, b.queryString)
+		}
```
Comment (Bobgy) [inline v2/component/launcher.go:81]:
```suggestion
	// append prefix=b.prefix to existing queryString
	q := b.queryString
	if len(b.prefix) > 0 {
		if len(q) > 0 {
			q = q + "&prefix=" + b.prefix
		} else {
			q = "?prefix=" + b.prefix
		}
	}
	u = u + q
```

Do you think this gets easier to read?

**Review by Bobgy — COMMENTED** [review]
/lgtm
Great work!
I've only got some nit pickings

File: sdk/python/kfp/compiler/_default_transformers.py [hunk 2]
```diff
@@ -52,5 +52,21 @@ def add_kfp_pod_env(op: BaseOp) -> BaseOp:
...
+              optional=True
+      )))
+  ).add_env_variable(
+      k8s_client.V1EnvVar(
+          name = 'AWS_SECRET_ACCESS_KEY',
+          value_from=k8s_client.V1EnvVarSource(secret_key_ref=k8s_client.V1SecretKeySelector(
+              name='kfp-s3-secret',
+              key='secretkey',
+              optional=True
+          ))))
```
[the remark this answers is not in the document]
Comment (capri-xiyue) [reply sdk/python/kfp/compiler/_default_transformers.py:71]: Sure. Can we do it in another PR? Otherwise, this PR may get too large.

File: samples/test/minio_pipeline_root_two_step_test.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+def verify(run, run_id: str):
+    assert run.status == 'Succeeded'
+    # TODO(Bobgy): verify MLMD status
+
+
+run_pipeline_func([
+    TestCase(pipeline_func=minio_pipeline_root_two_step_pipeline,
+             arguments={
+                 kfp.dsl.ROOT_PARAMETER_NAME: 's3://mlpipeline/output-directory/v2-artifacts?region=minio&endpoint=minio-service.kubeflow:9000&disableSSL=true&s3ForcePathStyle=true'},
```
[the remark this answers is not in the document]
Comment (capri-xiyue) [reply samples/test/minio_pipeline_root_two_step_test.py:30]: Amazon S3 supports both virtual-hosted–style and path-style URLs to access a bucket. FYI, https://github.com/aws/aws-sdk-go/blob/main/aws/config.go#L118. I followed the example of https://gocloud.dev/howto/blob/#s3-compatible

**Comment by capri-xiyue** [comment]
> users have multiple different minio instances



> I just checked the run result in test infra, looks great!
> https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/6842e5e8-df1b-456b-a358-c74571007cdf
> 
> Would URIs be more informative if we keep the parameters including endpoint? e.g. if users have multiple different minio instances, they can use the endpoint to distinguish minio URIs. I think for stuff we store in MLMD, we want them to be globally unique identifiers.
> image
> 
> I think a URL without the querystring might be easier to read, but we can choose to hide querystring in UI, but keep it in MLMD to reserve full information. What do you think?

I'm not sure whether it's common that user will use multiple different minio instances at the same time or user will change their minio instances when they use KFP. I checked the KFP V1. In KFP V1, we don't different the URL with the endpoint
KFP V1 artifacts.  I think currently for KFP V2 compatibility, it's fine that we keep the same behavior as that in V1. Later, after we get more feedback from users, we can change the implementation.
I'd prefer the MLMD and the UI show the same info to avoid confusion.

File: v2/component/launcher.go (hunk 1 above)
Comment (capri-xiyue) [reply v2/component/launcher.go]: done

File: samples/test/minio_pipeline_root_two_step_test.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+             verify_func=verify)
+])
```
[the remark this answers is not in the document]
Comment (capri-xiyue) [reply samples/test/minio_pipeline_root_two_step_test.py]: done

**Comment by capri-xiyue** [comment]
Minio Test passed https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/6f92b7b0-8cae-4343-8033-18e908aeffb0

File: v2/component/launcher.go [hunk 5]
```diff
@@ -697,3 +709,21 @@ func getExecutorOutput() (*pipeline_spec.ExecutorOutput, error) {
 
 	return executorOutput, nil
 }
+
+func (l *Launcher) openBucket() (*blob.Bucket, error) {
+	if l.bucketConfig.scheme == "minio://" {
+		sess, err := session.NewSession(&aws.Config{
+			Credentials:      credentials.NewStaticCredentials("minio", "minio123", ""),
+			Region:           aws.String("minio"),
+			Endpoint:         aws.String("minio-service.kubeflow:9000"),
```
Comment (Bobgy) [inline v2/component/launcher.go:718]: this is tricky, namespace can be different in different envs:

For KFP standalone & full Kubeflow, the default is `kubeflow` namespace.
However, for AI Platform Pipelines, the default is `default` namespace and it's very easy for users to choose any namespace.

So we have a few options:
1. use endpoint `minio-service:9000`, this assumes minio service is in the same namespace, so it works for KFP standalone and AI Platform Pipelines, but not for full Kubeflow.
2. use endpoint `minio-service.kubeflow:9000`, this works for KFP standalone and full Kubeflow, but not AI Platform Pipelines

If we choose one of the two, 1. is better now, because next full Kubeflow release is at least 3 months from now, so we have some time to prepare for it.

Ideally, this should be configurable using a configmap + a secret. We can add this consideration for full artifact repository support after v2 compatible mode MVP.

File: samples/test/minio_pipeline_root_two_step_test.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Two step v2-compatible pipeline with minio pipeline_root."""
+
+import kfp
+
+from .two_step import two_step_pipeline
```
Comment (Bobgy) [inline samples/test/minio_pipeline_root_two_step_test.py:18]: this doesn't need to be a separate test file, you can put this in the same file just as a separate TestCase

File: sdk/python/kfp/compiler/_default_transformers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bobgy) [reply sdk/python/kfp/compiler/_default_transformers.py]: SGTM, there's no need to verify this now

**Comment by Bobgy** [comment]
/lgtm
I think this mostly looks good to me, great work!
We can merge once immediate comments are resolved

File: v2/component/launcher.go (hunk 5 above)
Comment (capri-xiyue) [reply v2/component/launcher.go:718]: done

File: samples/test/minio_pipeline_root_two_step_test.py (hunk 6 above)
Comment (capri-xiyue) [reply samples/test/minio_pipeline_root_two_step_test.py:18]: done

File: v2/cmd/launch/main.go [hunk 7]
```diff
@@ -30,7 +30,7 @@ var (
 	pipelineName      = flag.String("pipeline_name", "", "The current pipeline name.")
 	pipelineRunID     = flag.String("pipeline_run_id", "", "The current pipeline run ID.")
 	pipelineTaskID    = flag.String("pipeline_task_id", "", "The current pipeline task ID.")
-	pipelineRoot      = flag.String("pipeline_root", "", "The root output directory in which to store output artifacts.")
+	pipelineRoot      = flag.String("pipeline_root", "minio://mlpipeline/artifacts", "The root output directory in which to store output artifacts.")
```
[the remark this answers is not in the document]
Comment (capri-xiyue) [reply v2/cmd/launch/main.go:33]: done

**Comment by capri-xiyue** [comment]
e2e test passed https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/478963fc-fb5d-4c98-ac05-c10f3e10cf6b

File: v2/component/launcher.go [hunk 8]
```diff
@@ -697,3 +709,21 @@ func getExecutorOutput() (*pipeline_spec.ExecutorOutput, error) {
 
 	return executorOutput, nil
 }
+
+func (l *Launcher) openBucket() (*blob.Bucket, error) {
+	if l.bucketConfig.scheme == "minio://" {
+		sess, err := session.NewSession(&aws.Config{
+			Credentials:      credentials.NewStaticCredentials("minio", "minio123", ""),
```
Comment (Bobgy) [inline v2/component/launcher.go:716]: As discussed, I believe we should get the secret from [mlpipeline-minio-artifact-secret](https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/third-party/minio/base/mlpipeline-minio-artifact-secret.yaml). Do you plan to do it in a separate PR or in this PR? I think both works

**Review by Bobgy — COMMENTED** [review]
Looks great!
/lgtm
only confirming your plan for one thing

File: v2/component/launcher.go (hunk 8 above)
Comment (capri-xiyue) [reply v2/component/launcher.go:716]: Will do it in another PR
